### PR TITLE
Ejudge integration

### PR DESCRIPTION
**Pull request by perveevm**

### CODE REVIEW

File: overlay/package-lock.json [hunk 1]
```diff
@@ -4437,6 +4437,17 @@
         "url": "https://github.com/sponsors/sindresorhus"
       }
     },
+    "node_modules/ansi-escapes/node_modules/type-fest": {
```
Comment (kunyavskiy) [inline overlay/package-lock.json:4440]: Is this change intentional? @irdkwmnsb can you check if it is ok?

File: backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+            "PR" to "",
+            "RJ" to "",
+            "SK" to "",
+            "SY" to "",
+            "SM" to "",
+            "RU" to "",
+            "CD" to "",
+            "CG" to "",
+            "AV" to "",
+            "RJ" to "",
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt:275]: Looks RJ is used twice in the list

File: backend/config/archive/nac22practice4/events.properties [hunk 3]
```diff
@@ -1,4 +1,4 @@
 login=$creds.kattis_login
 password=$creds.kattis_token
 url=https://open.kattis.com/clics-api/contests/nac22practice4
-standings.type=CLICS
+standings.type=CLICS
```
Comment (kunyavskiy) [inline backend/config/archive/nac22practice4/events.properties:4]: this one looks unintentional too

File: backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+            ContestStatus.BEFORE
+        } else {
+            ContestStatus.RUNNING
+        }
+
+        if (status == ContestStatus.RUNNING && contestData.status !== ContestStatus.RUNNING) {
+            contestData.startTime = startTime
+        }
+        contestData.status = status
+        contestData.contestTime = Duration.between(
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt:152]: what is wrong with currentTime - startTime? As I remember it returns kotlin.time.Duration?

File: backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+            parseRunInfo(contestInfo, run, onRunChanges)
+        }
+    }
+
+    private fun parseRunInfo(
+        contestInfo: EjudgeContestInfo,
+        element: Element,
+        onRunChanges: (RunInfo) -> Unit
+    ) {
+        val time = (element.attr("time").toLong() * 1000).milliseconds
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt:182]: .seconds?

File: backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+        val run = RunInfo(
+            id = oldRun?.id ?: element.attr("run_id").toInt(),
+            isAccepted = "AC" == result,
+            isJudged = "" != result,
+            isAddingPenalty = "AC" != result && "CE" != result,
+            result = result,
+            problemId = problemId,
+            teamId = teamId,
+            percentage = percentage,
+            time = time.inWholeMilliseconds.toDuration(DurationUnit.MILLISECONDS),
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt:214]: What's the purpose of converting to int and back?

File: backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+    init {
+        val problemsInfo = parseProblemsInfo()
+        val teamsInfo = parseTeamsInfo(problemsInfo.size)
+
+        contestData = EjudgeContestInfo(
+            problemsInfo,
+            teamsInfo.associateBy { it.teamInfo.contestSystemId },
+            Instant.fromEpochMilliseconds(0),
+            ContestStatus.UNKNOWN,
+            properties.getProperty("contest.length")?.toInt()?.milliseconds ?: 5.hours,
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt:241]: maybe this data exists in external.xml? It would be much better to retrieve it from it. In PCMS it's a hack, which is required due to lack of information

File: backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+        val problemId = element.attr("prob_id").toInt() - 1
+        val isFrozen = time >= contestInfo.freezeTime
+        val oldRun = contestInfo.teams[teamSystemId.toString()]!!.runs[problemId].getOrDefault(runId, null)
+        val result = when {
+            isFrozen -> ""
+            else -> statusMap.getOrDefault(element.attr("status"), "WA")
+        }
+        val percentage = when {
+            isFrozen -> 0.0
+            "" == result -> 0.0
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt:189]: I see a test field in xml? Does it exist when the run is in progress?

Comment (perveevm) [reply backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt:189]: Either it exists, or it equals to 0 when the run is in progress, I don't remember. But both of this cases are useless for this functionality.
